x/tally/keeper/gas_meter.go

```diff
@@ -0,0 +1,190 @@
+
+	return dists
+}
+
+// MeterProxyGas computes and records the gas consumption of data proxies given
+// proxy public keys in basic consensus and the request's replication factor.
+func (k Keeper) MeterProxyGas(ctx sdk.Context, proxyPubKeys []string, replicationFactor uint16, gasMeter *types.GasMeter) {
+	if len(proxyPubKeys) == 0 || gasMeter.RemainingExecGas() == 0 {
+		return
+	}
+
+	for _, pubKey := range proxyPubKeys {
+		pubKeyBytes, err := hex.DecodeString(pubKey)
+		if err != nil {
+			k.Logger(ctx).Error("failed to decode proxy public key", "error", err, "public_key", pubKey)
+			continue
+		}
+		proxyConfig, err := k.dataProxyKeeper.GetDataProxyConfig(ctx, pubKeyBytes)
+		if err != nil {
+			k.Logger(ctx).Error("failed to get proxy config", "error", err, "public_key", pubKey)
+			continue
+		}
+
+		gasUsedPerExec := proxyConfig.Fee.Amount.Quo(gasMeter.GasPrice()).Uint64()
+		gasUsedPerExec = min(gasUsedPerExec, gasMeter.RemainingExecGas()/uint64(replicationFactor))
+		gasUsed := gasUsedPerExec * uint64(replicationFactor)
+
+		gasMeter.ConsumeExecGasForProxy(pubKeyBytes, gasUsed)
+	}
+}
+
+// MeterExecutorGasFallback computes and records the gas consumption of committers
+// of a data request when basic consensus has not been reached. If checkReveal is
+// set to true, it will only consume gas for committers that have also revealed.
+func MeterExecutorGasFallback(req types.Request, gasCostCommit uint64, gasMeter *types.GasMeter, checkReveal bool) {
+	if len(req.Commits) == 0 || gasMeter.RemainingExecGas() == 0 {
+		return
+	}
+	if checkReveal && len(req.Reveals) == 0 {
+		return
+	}
+
+	i := 0
+	committers := make([]string, len(req.Commits))
+	for k := range req.Commits {
+		committers[i] = k
+		i++
+	}
+	sort.Strings(committers)
+
+	var gasLimitPerExec uint64
+	if checkReveal {
+		gasLimitPerExec = gasMeter.RemainingExecGas() / uint64(len(req.Reveals))
+	} else {
+		gasLimitPerExec = gasMeter.RemainingExecGas() / uint64(len(committers))
+	}
+
+	for _, committer := range committers {
+		if checkReveal {
+			if _, ok := req.Reveals[committer]; !ok {
+				continue
+			}
+		}
+		gasUsed := min(gasLimitPerExec, gasCostCommit)
+		gasMeter.ConsumeExecGasForExecutor(committer, gasUsed)
+	}
+}
+
+// MeterExecutorGasUniform computes and records the gas consumption of executors
+// when their gas reports are uniformly at "gasReport".
+func MeterExecutorGasUniform(executors []string, gasReport uint64, replicationFactor uint16, gasMeter *types.GasMeter) {
+	gasUsed := min(gasReport, gasMeter.RemainingExecGas()/uint64(replicationFactor))
+	for _, executor := range executors {
+		gasMeter.ConsumeExecGasForExecutor(executor, gasUsed)
+	}
+}
+
+// MeterExecutorGasDivergent computes and records the gas consumption of executors
+// when their gas reports are divergent.
+func MeterExecutorGasDivergent(executors []string, gasReports []uint64, replicationFactor uint16, gasMeter *types.GasMeter) {
```

Comment:
I think these 2 functions should consider the total proxy gas consumed when calculating the executor gas used.

Let assume a uniform reporting scenario:
```
replication_factor: 10
exec_gas_limit: 1000
reported_gas: 9
data_proxy_gas_cost: 5 -> data_proxy_gas_consumed: 50
```

The current implementation would give us the following executor gas used:
```
gasUsed := min(gasReport, (exec_gas_limit - data_proxy_gas_consumed) / replication_factor))
gasUsed := min(9, (1000 - 50) / 10)
gasUsed := min (9, 95)
gasUsed := 9
```

While it should be:
```
gasUsed := min(gasReport - data_proxy_gas_cost, (exec_gas_limit - data_proxy_gas_consumed) / replication_factor))
gasUsed := min(9 - 5, (1000 - 50) / 10))
gasUsed := min(4, 95)
gasUsed := 4
```

Reply:
I see where I went wrong, we do the adjustment elsewhere. I still changed it since right now the implementation relies on no other operations having changed the exec gas which feels a little brittle. Obviously my change also relies on the data proxy calculation to have been done first, but it won't break if we introduce some other tweak to the gas calculation.